toolchain/sem_ir/builtin_function_kind.cpp

```diff
@@ -0,0 +1,177 @@
+// ValidateSignature for details.
+struct T {
+  static auto Check(const File& /*sem_ir*/, ValidateState& state,
+                    TypeId type_id) -> bool {
+    if (state.t.is_valid() && type_id != state.t) {
+      return false;
+    }
+    state.t = type_id;
+    return true;
+  }
+  static auto Get(ValidateState& state) -> TypeId { return state.t; }
+};
+
+// Constraint that requires the type to be an integer type. See
+// ValidateSignature for details.
+struct AnyInt {
+  static auto Check(const File& sem_ir, ValidateState& /*state*/,
+                    TypeId type_id) -> bool {
+    if (sem_ir.types().GetInstId(type_id) == InstId::BuiltinIntType) {
+      return true;
+    }
+    // TODO: Support iN for all N, and the Core.BigInt type we use to implement
+    // for integer literals.
+    return false;
+  }
+};
+}  // namespace
+
+// Validates that this builtin has a signature matching the specified signature.
+//
+// `SignatureFnType` is a C++ function type that describes the signature that is
+// expected for this builtin. For example, `auto (AnyInt, T) -> T`. Types used
+// within the signature should provide a `Check` function that validates that
+// the Carbon type is expected:
+//
+//   auto Check(const File&, ValidateState&, TypeId) -> bool;
+//
+// `Params` describes any generic parameters that appear in the signature, and
+// each listed type should be a `Param` specialization. For example,
+// `Param<T, AnyInt>` means `T:! AnyInt`.
+template <typename SignatureFnType, typename... Params>
+static auto ValidateSignature(const File& sem_ir,
+                              llvm::ArrayRef<TypeId> arg_types,
+                              TypeId return_type) -> bool {
+  using SignatureTraits = llvm::function_traits<SignatureFnType*>;
+  ValidateState state;
+
+  // Must have expected number of arguments.
+  if (arg_types.size() != SignatureTraits::num_args) {
+    return false;
+  }
+
+  // Argument types must match.
+  if (![&]<std::size_t... Indexes>(std::index_sequence<Indexes...>) {
+        return ((SignatureTraits::template arg_t<Indexes>::Check(
+                    sem_ir, state, arg_types[Indexes])) &&
+                ...);
+      }(std::make_index_sequence<SignatureTraits::num_args>())) {
+    return false;
+  }
+
+  // Result type must match.
+  if (!SignatureTraits::result_t::Check(sem_ir, state, return_type)) {
+    return false;
+  }
+
+  // Generic parameters must have the right types.
+  if (!((Params::Check(sem_ir, state)) && ...)) {
+    return false;
+  }
+  return true;
+}
+
+// Descriptions of builtin functions follow. For each builtin, a corresponding
+// `BuiltinInfo` constant is declared describing properties of that builtin.
+
+// Not a builtin function.
+constexpr BuiltinInfo None = {"", nullptr};
+
+// "int.add": integer addition.
+constexpr BuiltinInfo IntAdd = {
+    "int.add", ValidateSignature<auto(T, T)->T, Param<T, AnyInt>>};
```

Comment:
I'm confused by this, maybe you could explain directly what this example is doing in ValidateSignature's comment?

Reply:
I've expanded the comment to walk through a similar but slightly more interesting example.